misc/release_test.sh

```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env bash
+
+# Taichi release test suite
+
+# This script is created mainly for eyeball-testing
+# that if all of the official examples are still working
+# with the latest version of Taichi.
+
+# The test cases are created from external repositories
+# please reach out to us if you are the owner of those
+# repos and don't like us to do it.
+
+# You can add more tests into this script and plug-n-play
+# existing tests in the `taichi::test::main` function as
+# you need.
+
+function taichi::utils::set_debug {
+    if [ ${DEBUG} == "true" ]; then
+        set -x
+    fi
+    set -euo pipefail
+}
+
+function taichi::utils::logger {
+    # default: gray
+    if [ "$1" == "info" ]; then
+        printf '\e[1;90m%-6s\e[m\n' "$(date +"[%m-%d %H:%M:%S]") $2"
+    # error: red
+    elif [ "$1" == "error" ]; then
+        printf '\e[1;91m%-6s\e[m\n' "$(date +"[%m-%d %H:%M:%S]") $2"
+    # success: green
+    elif [ "$1" == "success" ]; then
+        printf '\e[1;92m%-6s\e[m\n' "$(date +"[%m-%d %H:%M:%S]") $2"
+    # warning: yellow
+    elif [ "$1" == "warning" ]; then
+        printf '\e[1;93m%-6s\e[m\n' "$(date +"[%m-%d %H:%M:%S]") $2"
+    # debug: gray
+    elif [ "$1" == "debug" ]; then
+        if [ "${DEBUG}" == "true" ]; then
+            printf '\e[1;90m%-6s\e[m\n' "$(date +"[%m-%d %H:%M:%S]") $2"
+        fi
+    else
+        printf "$1"
+    fi
+}
+
+function taichi::utils::logger::info {
+    taichi::utils::logger "info" "$1"
+}
+
+function taichi::utils::logger::error {
+    taichi::utils::logger "error" "$1"
+}
+
+function taichi::utils::logger::success {
+    taichi::utils::logger "success" "$1"
+}
+
+function taichi::utils::logger::warning {
+    taichi::utils::logger "warning" "$1"
+}
+
+function taichi::utils::logger::debug {
+    taichi::utils::logger "debug" "$1"
+}
+
+function taichi::utils::line {
+    printf '%.0s-' {1..20}; echo
+}
+
+function taichi::utils::git_clone {
+    local GIT_ORG=$1
+    local GIT_REPO=$2
+    git clone "git@github.com:${GIT_ORG}/${GIT_REPO}.git"
+}
+
+function taichi::utils::pause {
+    read -p "Press enter to continue"
+}
+
+function taichi::test::ggui {
+    local WORKDIR=${1}
+    local PATTERN="*_ggui.py"
+    local ORG="taichi-dev"
+    local REPO="taichi"
+
+    # divider
+    taichi::utils::line
+    taichi::utils::logger::info "Running GGUI examples"
+    
+    # clone the repo
+    taichi::utils::git_clone "${ORG}" "${REPO}"
+    cd "${REPO}/python/taichi/examples/ggui_examples"
```

Comment:
I think we could leverage `pushd/popd` to switch working dir here, i.e.:
```bash
pushd  "${REPO}/python/taichi/examples/ggui_examples"

# run tests
...

popd
```

and we don't need to pass current working dir if we use `pushd/popd`.

Reply:
> pushd/popd aren't portable

The shebang at the beginning is `#!/usr/bin/env bash`, I thought we would always use bash.

> it would change the dirs stack and require us to use +N in the script

I am curious about the reason that we should concert about the dirs stack though. I think we don't need to specify `+N` neither.. For me, `pushd/popd` are convenient commands which temporarily change working dir.

This is not a blocker, the current script is great 👍 !

Reply:
> The shebang at the beginning is #!/usr/bin/env bash, I thought we would always use bash.

You are right, I forgot about shebang :P

>  I am curious about the reason that we should concert about the dirs stack though.

It's usually cauz we have to manually `cd`/`pushd` within the dirs for each test block in this script after cloning the repo as we don't know what dependencies we need to install in some external repos.

Reply:
> we have to manually cd/pushd within the dirs for each test block in this script after cloning the repo

hmmm but we would call `popd` at the end of each test block, so I think the dirs stack would remain the same after each test block.

Reply:
per offline discussion, cd changes the stack of `dirs`, so I'm going to keep this approach for now :P